File: go/300.500/430.FlattenAMultilevelDoublyLinkedList.go
package main

import "fmt"

type Node struct {
	Val   int
	Prev  *Node
	Next  *Node
	Child *Node
}

func flatten(root *Node) *Node {
	ret := root
	var dfs func(root *Node) *Node
	dfs = func(root *Node) *Node {
		for root != nil {
			if root.Child == nil {
				if root.Next == nil {
					return root
				} else {
					root = root.Next
				}
			} else {
				next := root.Next
				root.Next = root.Child
				root.Child.Prev = root
				root.Child = nil
				root = dfs(root.Next)
				if next != nil {
					root.Next = next
					next.Prev = root
					root = next
				}
			}
		}
		return root
	}
	dfs(root)
	return ret
}

func print(root *Node) {
	for root != nil {
		c := root
		root = root.Next
		fmt.Printf("%d,", c.Val)
	}
	fmt.Println()
}

func main() {
	head := &Node{Val: 1}
	h := head
	head.Next = &Node{Val: 2}
	head.Next.Prev = head
	head = head.Next
	head.Next = &Node{Val: 3}
	head.Next.Prev = head
	head = head.Next
	head.Next = &Node{Val: 4}
	head.Next.Prev = head
	head = head.Next
	head.Next = &Node{Val: 5}
	head.Next.Prev = head
	head = head.Next
	head.Next = &Node{Val: 6}
	head.Next.Prev = head
	head = h.Next.Next
	head.Child = &Node{Val: 7}
	head = head.Child
	head.Next = &Node{Val: 8}
	head.Next.Prev = head
	head = head.Next
	head.Next = &Node{Val: 9}
	head.Next.Prev = head
	head = head.Next
	head.Next = &Node{Val: 10}
	head.Next.Prev = head
	head = h.Next.Next.Child.Next
	head.Child = &Node{Val: 11}
	head = head.Child
	head.Next = &Node{Val: 12}
	head.Next.Prev = head
	ret := flatten(h)
	print(ret)
}

File: 300.500/430.FlattenAMultilevelDoublyLinkedList.cpp
#include <iostream>
using namespace std;

class Node {
public:
    int val{};
    Node* prev{};
    Node* next{};
    Node* child{};
    Node(int i) :val(i) {}
};

class FlattenAMultilevelDoublyLinkedList{
public:
    Node* flatten(Node* head) {
        ret = head;
        dfs(head);
        return ret;
    }
private:
    Node* dfs(Node* head) {
        while (head) {
            if (head->child == nullptr) {
                if (head->next)
                    head = head->next;
                else
                    return head;
            }
            else {
                Node* next = head->next;
                head->next = head->child;
                head->child->prev = head;
                head->child = nullptr;
                head = dfs(head->next);
                if (next) {
                    head->next = next;
                    next->prev = head;
                    head = head->next;
                }
            }
        }
        return head;
    }
private:
    Node* ret;
};

void print(Node* head) {
    cout << "[";
    while (head)
    {
        Node* c = head;
        head = head->next;
        cout << c->val << ",";
        delete c;
    }
    cout << "]" << endl; 
}

int main(int argc, char const* argv[]) {
    FlattenAMultilevelDoublyLinkedList cls;
    Node* head = new Node(1);
    Node* h = head;
    head->next = new Node(2);
    head->next->prev = head;
    head = head->next;
    head->next = new Node(3);
    head->next->prev = head;
    head = head->next;
    head->next = new Node(4);
    head->next->prev = head;
    head = head->next;
    head->next = new Node(5);
    head->next->prev = head;
    head = head->next;
    head->next = new Node(6);
    head->next->prev = head;
    head = h->next->next;
    head->child = new Node(7);
    head = head->child;
    head->next = new Node(8);
    head->next->prev = head;
    head = head->next;
    head->next = new Node(9);
    head->next->prev = head;
    head = head->next;
    head->next = new Node(10);
    head->next->prev = head;
    head = h->next->next->child->next;
    head->child = new Node(11);
    head = head->child;
    head->next = new Node(12);
    head->next->prev = head; 
    Node* r = cls.flatten(h);
    print(r);
    return 0;
}
